generate/generate.go

```diff
@@ -134,19 +134,91 @@ func (g *Generator) GeneratePkg(path string, gens []config.Generator) error {
 	}
 	// Run any other configured generators.
 	for _, gen := range gens {
-		if err := g.generatePlugin(*req, gen); err != nil {
-			return err
+		if gen.IsProtoc() {
+			if err := g.generateProtoc(*req, gen); err != nil {
+				return err
+			}
+		} else {
+			if err := g.generatePlugin(*req, gen); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
-func (g *Generator) generatePlugin(req plugin.CodeGeneratorRequest, gen config.Generator) error {
-	protoParameters := make([]string, len(gen.Params))
-	for i, p := range gen.Params {
-		protoParameters[i] = fmt.Sprintf("%s=%s", p.Key, p.Value)
+func (g *Generator) generateProtoc(req plugin.CodeGeneratorRequest, gen config.Generator) error {
+	// The proto files we are asking protoc to generate. This should be
+	// a formatted list of what is in req.FileToGenerate.
+	protoFilenames := []string{}
+
+	fds := desc.FileDescriptorSet{
+		File: req.ProtoFile,
+	}
+	// Determine the files that protoc will be generating and
+	// remove the package directory path and change .gunk to .proto.
+	// This is needed because there is no way to specify the output
+	// filename, and protoc will use the FileDescriptorSet.File.Name
+	// to determine the filename. In our case, that would be the
+	// package path, but it would be created from the current relative
+	// directory, rather than an absolute path.
+	//
+	// We also replace .gunk with .proto. protoc will turn
+	// 'echo.gunk' into 'echo.gunk_pb.js' which makes it a bit
+	// hard to replace afterwards. This seemed like the easier approach.
+	for i, f := range fds.File {
+		for _, ftg := range req.GetFileToGenerate() {
+			if f.GetName() == ftg {
+				_, name := filepath.Split(ftg)
+				name = strings.Replace(name, ".gunk", ".proto", 1)
+				fds.File[i].Name = proto.String(name)
+				protoFilenames = append(protoFilenames, name)
+			}
+		}
+	}
+
+	bs, err := proto.Marshal(&fds)
+	if err != nil {
+		return err
+	}
+
+	// Create a temporary file to store the proto filedescriptorset.
+	tmpfile, err := ioutil.TempFile("", "gunk.bin")
+	if err != nil {
+		return fmt.Errorf("unable to open tmp file: %v", err)
+	}
+	protoFileDescriptorPath := tmpfile.Name()
+	defer os.Remove(protoFileDescriptorPath)
+
+	if _, err := tmpfile.Write(bs); err != nil {
+		return fmt.Errorf("unable to write to tmp file: %v", err)
 	}
-	req.Parameter = proto.String(strings.Join(protoParameters, ","))
+	if err := tmpfile.Close(); err != nil {
+		return fmt.Errorf("unable to close tmp file: %v", err)
+	}
+
+	// Build up the protoc command line arguments.
+	command := "protoc"
+	args := []string{
+		fmt.Sprintf("--%s_out=%s", gen.ProtocGen, gen.ParamsWithOut()),
+		fmt.Sprintf("--descriptor_set_in=%s", protoFileDescriptorPath),
```

Comment:
I'd use `/dev/stdin` instead. It's not portable to Windows, but we don't care about that for now. Also, we use `/dev/stdout` elsewhere :)

I did report thiss to upstream ages ago, but unfortunately it has had zero attention so far: https://github.com/protocolbuffers/protobuf/issues/4163

Reply:
Is there a benefit to using stdin here? We create and delete temp files when using protoc to load the imports, and as you said, it may make it _possibly_ windows compatible.

Reply:
Both `/dev/` and temporary files are hacks - I'm simply saying we should be consistent. Elsewhere in the code we already use `--descriptor_set_out=/dev/stdout`, so I think you should do the equivalent here.

Unless you prefer to use temporary files in both. We don't even test on Windows, so I personally think it's not a priority to guesstimate what will work on Windows.

Reply:
Yup, makes sense. I have updated it to use /dev/stdin.